### DISCUSSION TITLE
Access zod IO schema in middleware

### DISCUSSION THREAD

**Discussion by noahdamiani** [RobinTail/express-zod-api#1066] [Q&A] [answered]
Is it possible to access the IO schema as a callback parameter in middleware? I scrolled through all the properties of the request, response, and input params given by the middleware callback but did not see anything schema related, hoping I missed something. Outlining what I am trying to do below:

```
const middleware = createMiddleware({
  input: z.object({}),
  async middleware({ request, response }) {
    // Middleware is attached to each route, so in theory it would be possible 
    // to resolve the input/output schema for a given route?
    // i.e. request.route.input // request.route.output

    return doStuffWithZodModel(request.route.input)
  },
});

const withSchemaMiddleware = defaultEndpointsFactory.addMiddleware(middleware);

const helloWorldEndpoint = withSchemaMiddleware.build({
     ...
});
```

**Comment by RobinTail** [comment, marked as answer]
Hello @noahdamiani ,

I don't understand the use-case.
What are you trying to achieve exactly?

Handlers of the both `Middleware` and `Endpoint` do not receive the schema — they receive the parsed and validated input data.

However, it's easy for the handler to be aware of the schema — just declare it in a const in outer scope.

```ts

const mySchema = z.object({ /* ... */ });

const myMiddleware = createMiddleware({
  input: mySchema,
  middleware: async () => {
    return doYourStuff( mySchema )
  }
})
```

**Reply by noahdamiani** [reply]
I'm hoping to use the routes config as a mapping to retrieve a request schema given an endpoint uri. The example you've provided defines a model explicitly, whereas I am trying to retrieve the model dynamically. It sounds like middleware is not aware of the endpoint object or the uri provided in the routes config - I was hoping to access those in the middleware callback.

**Reply by RobinTail** [reply]
@noahdamiani , `zod` is a static type validator. The input schema assigned to the middleware or endpoint can not be dynamic. It can only be loose enough to fit all the variations, for example using `.or()` or using some refinement, based on the data being parsed.
